feat: Add JSON/YAML config formats and environment variable support

eminwux opened the pull request:
## Summary

This PR adds YAML configuration format support and full environment variable configuration alongside the existing HCL format. Closes #223.

## Problem

- Containerized deployments (Kubernetes, Docker) need environment-variable-driven configuration.
- Users requested YAML as a modern, human-friendly config format.
- HCL already supports its JSON representation natively, but YAML requires explicit support.

## Solution

### 1. YAML configuration support

- Added `.yaml` / `.yml` parsing alongside HCL's `.conf` / `.json`.
- Automatic format detection from the file extension, or explicit selection via the `-config-format` flag (`hcl` | `yaml` | `auto`).
- Unknown-key detection for YAML, matching HCL's strict validation.

### 2. Environment variable configuration

- All configuration fields can be set via environment variables with the `SPIFFE_HLP_` prefix.
- Environment variables override values from a YAML config file when both are provided.
- Configuration can be provided entirely via environment variables when no config file is specified or the file does not exist.

### 3. Custom parsing for complex types

- Indexed environment variables for `jwt_svids` arrays, since `cleanenv` does not support arrays of structs:
  - `SPIFFE_HLP_JWT_SVIDS="2"` — count-based (uses indices 0, 1)
  - `SPIFFE_HLP_JWT_SVIDS="0,2"` — explicit indices
  - Per-index fields: `SPIFFE_HLP_JWT_SVIDS_0_AUDIENCE`, `SPIFFE_HLP_JWT_SVIDS_0_SVID_FILE_NAME`, etc.
- Custom parsing for `DaemonMode` (`*bool`) because `cleanenv` does not support pointers to boolean values.

### 4. Debug logging

Debug-level log of the fully reconciled configuration as comma-separated `key=value` pairs, emitted only when `log_level` is `debug`.

## A note on JSON

HCL already parses its [JSON representation](https://github.com/hashicorp/hcl#syntax) natively, so JSON configuration files are supported without any separate code path. `.json` files are routed through the HCL parser by auto-detection.

## HCL is not deprecated

An earlier revision of this PR proposed deprecating HCL; that was reverted after review feedback. HCL remains a first-class format across the SPIFFE org.

## Changes

- `cmd/spiffe-helper/config/config.go`: YAML parsing, env-var wiring, unknown-key detection, single file-existence check at the entry point.
- `cmd/spiffe-helper/config/config_env.go`: custom env-var parsing for `JWTSVIDs` and `DaemonMode`.
- `cmd/spiffe-helper/main.go`: log-level bootstrap from env var, cleaned up duplicate stat checks.
- `pkg/health/health.go`: `env` tags for health-check fields.
- `cmd/spiffe-helper/config/config_test.go`: unit tests for YAML parsing, env-only mode, env overrides, indexed `jwt_svids`, unknown keys.
- `script/docker/test-env-vars.sh` + `Makefile`: Docker-based integration tests exercising real env-var configuration (`make test-integration-docker`).
- `README.md`: YAML / env-var documentation, indexed `jwt_svids` guide.

## Examples

### YAML

```yaml
agent_address: "/tmp/spire-agent/public/api.sock"
cert_dir: "/opt/spiffe-helper/certs"
svid_file_name: "svid.pem"
svid_key_file_name: "svid_key.pem"
svid_bundle_file_name: "bundle.pem"
```

### Environment variables only

```bash
export SPIFFE_HLP_AGENT_ADDRESS="/tmp/agent.sock"
export SPIFFE_HLP_CERT_DIR="/opt/certs"
export SPIFFE_HLP_SVID_FILE_NAME="svid.pem"
export SPIFFE_HLP_SVID_KEY_FILE_NAME="svid_key.pem"
export SPIFFE_HLP_SVID_BUNDLE_FILE_NAME="bundle.pem"
spiffe-helper   # no config file needed
```

## Testing

- ✅ `go test ./...` — unit tests pass (YAML parsing, unknown keys, env-only mode, env overrides, indexed `JWTSVIDs`, `DaemonMode`, missing-required validation).
- ✅ `make test-integration-docker` — 10/10 Docker-based integration tests pass, verifying actual env-var configuration end-to-end.

## Backward compatibility

Fully backward compatible — no breaking changes:

- Existing HCL (`.conf`) and HCL-JSON (`.json`) files continue to work unchanged.
- Default behavior unchanged (still reads `helper.conf`).
- All existing functionality preserved.

eminwux committed:
feat: add JSON/YAML config and environment variable support

- Add JSON and YAML configuration format support (HCL deprecated)
- Add full environment variable configuration with SPIFFE_HLP_ prefix
- Add custom parsing for JWTSVIDs arrays and DaemonMode
- Add unknown key detection for JSON/YAML
- Add debug logging of reconciled configuration
- Add comprehensive unit and integration tests
- Update README with examples and documentation

Fully backward compatible - no breaking changes.

Signed-off-by: Emiliano Spinella <emilianofs@gmail.com>

faisal-memon commented on cmd/spiffe-helper/main.go:
```diff
@@ -21,23 +21,56 @@ const (
 	flag.Parse()
+
+	// Set initial log level from environment variable if provided (for early logging)
+	// This will be overridden by config value if set
```
Why is this backwards from other variables where env overrides config value?

eminwux committed:
fix: log level env var takes precedence over config

Signed-off-by: Emiliano Spinella <emilianofs@gmail.com>

faisal-memon commented on cmd/spiffe-helper/main.go:
```diff
@@ -32,21 +33,52 @@ func main() {
+	configFileToUse := *configFile
+	if _, err := os.Stat(*configFile); os.IsNotExist(err) {
+		log.Info("Configuration file not found, configuring via environment variables")
+		configFileToUse = ""
```
Similar check is done in `ParseConfigFile()`. Are both needed?

eminwux replied:
Fixed in 2a5598d. The `os.Stat` block in `main.go` is gone — the file-existence check now lives in one place at the top of `ParseConfigFile`, which falls back to env-only mode if the file is empty or missing.

faisal-memon commented on cmd/spiffe-helper/config/config.go:
```diff
@@ -28,57 +32,283 @@ const (
+			return nil, err
+		}
+		return config, nil
+	}
```
Duplicate check with `ProcessConfigFile()`

eminwux replied:
Fixed in 2a5598d. All three duplicate checks (`main.go`, `ParseAutoConfigFile`, and the one inside `ProcessConfigFileAndEnv`) are collapsed into a single guard at the top of `ParseConfigFile`. While I was in there I also renamed `ProcessConfigFileAndEnv` → `ParseYAMLConfigAndEnv` and removed the unused `ParseYAMLConfigFile` helper.

eminwux committed:
refactor(config): drop JSON tags and un-deprecate HCL per review feedback

Signed-off-by: Emiliano Spinella <emilianofs@gmail.com>

eminwux commented:
Done in 2a5598d.

- Dropped `json:` struct tags from `Config`, `JWTConfig`, and `health.Config` — now only `hcl` + `yaml`.
- Removed the `json` case from `-config-format` and the separate JSON parsing path. `.json` files now route through the HCL parser, which handles HCL's JSON representation transparently, so existing JSON configs keep working.
- Un-deprecated HCL: removed the runtime warning and all README banners/notes about HCL being removed in 0.11.0.
- Swapped the `github.com/stretchr/testify/assert/yaml` (testify-internal) import for `gopkg.in/yaml.v3`, which we still need for the unknown-key detection path on YAML files.
- Tests: removed `TestDetectsUnknownJSONConfig` (HCL path already covers it via `TestDetectsUnknownHCLConfig`); converted the JSON cases in `TestConfigFileWithEnvOverrides` to YAML.

Also rebased on top of main to pick up the recent dependency bumps.

faisal-memon commented:
@eminwux Thanks for checking in. hadn't heard from you in a while so I started addressing some of the issues in this branch: pr-335-takeover-codex. Please take a look.

faisal-memon committed:
Apply maintainer review updates for config support

faisal-memon commented:
Thanks again for the contributions here. I merged the changes from the takeover branch into this PR so we can keep the review and merge path here while preserving the original PR.

faisal-memon committed:
Fix lint issues in config tests

faisal-memon committed:
Mark yaml dependency as direct

faisal-memon committed:
Organize config functions by visibility

faisal-memon committed:
Simplify ParseConfig error handling

faisal-memon committed:
Unexport internal config file parsers

faisal-memon committed:
Make config flag override helper internal

faisal-memon committed:
Reorder config helpers below public API

faisal-memon committed:
Merge branch 'main' into feat/cleanenv

Signed-off-by: Faisal Memon <fymemon@yahoo.com>

MarcosDY commented on cmd/spiffe-helper/config/config.go:
```diff
@@ -28,70 +31,66 @@ const (
-	if err := hcl.Decode(config, string(dat)); err != nil {
-		return nil, err
+func ParseConfig(configFile string, configFormat string, daemonModeFlag bool, daemonModeFlagName string) (*Config, error) {
+	if configFile == "" {
```
since cmd has a default configfile it is never get into this if right?
and doc says it is going to use envvar when empty or file not found

faisal-memon replied:
Adjusted ParseConfig so a missing or empty config file now loads from env vars, which makes the documented env-only path reachable even with the default helper.conf value.

MarcosDY commented on cmd/spiffe-helper/config/config.go:
```diff
@@ -28,70 +31,66 @@ const (
+	// Note: DaemonMode does not have an env tag because cleanenv doesn't support *bool types.
+	// Instead, use populateDaemonModeFromEnv for environment variable support.
+	DaemonMode  *bool         `hcl:"daemon_mode" yaml:"daemon_mode"`
+	HealthCheck health.Config `hcl:"health_checks" yaml:"health_checks" env:"SPIFFE_HLP_HEALTH_CHECKS"`
```
the envvar here is not used since it is an object that contains it's own environment var right?

faisal-memon replied:
Updated this to use caarlos0/env nested prefixes instead of treating the whole health config as one env var. The parent now uses envPrefix:"HEALTH_" and the nested health fields define their suffixes.

MarcosDY commented on pkg/health/health.go:
```diff
@@ -14,10 +14,10 @@ import (
+	ListenerEnabled bool   `hcl:"listener_enabled" yaml:"listener_enabled" env:"SPIFFE_HLP_LISTENER_ENABLED"`
+	BindPort        int    `hcl:"bind_port" yaml:"bind_port" env:"SPIFFE_HLP_BIND_PORT"`
+	LivenessPath    string `hcl:"liveness_path" yaml:"liveness_path" env:"SPIFFE_HLP_LIVENESS_PATH"`
+	ReadinessPath   string `hcl:"readiness_path" yaml:"readiness_path" env:"SPIFFE_HLP_READINESS_PATH"`
```
may this ones contains SPIFFE_HLP_HEALTH_*?

faisal-memon replied:
Yes, changed these to SPIFFE_HLP_HEALTH_* and updated the tests/docs/script accordingly.

MarcosDY commented on README.md:
```diff
@@ -7,24 +7,27 @@
-The configuration file is an [HCL](https://github.com/hashicorp/hcl) formatted file that defines the following configurations:
+The configuration file can be in HCL, JSON, or YAML format. The format is automatically detected from the file extension (`.conf` for HCL, `.json` for JSON, `.yaml` or `.yml` for YAML), or can be explicitly specified using the `-config-format` flag.
+
+If the configuration file is not specified, does not exist, or is empty, all configuration can be provided via environment variables (see [Environment Variables](#environment-variables) section below).
```
may we contains the envvars on the table bellow?

faisal-memon replied:
Added the environment variable mapping table to the README, including the SPIFFE_HLP_HEALTH_* variables.

MarcosDY commented on cmd/spiffe-helper/config/config.go:
```diff
@@ -256,6 +231,149 @@ func NewSidecarConfig(config *Config, log logrus.FieldLogger) *sidecar.Config {
+	return err == nil
+}
+
+func detectConfigFormat(file string) (string, error) {
```
previous version fallback into hcl as default on files without extension, 
But now it is failing with invalid config file, may we use hcl as default?

faisal-memon replied:
Restored the backward-compatible behavior here: auto-detect now treats files without a recognized extension as HCL.

MarcosDY commented on cmd/spiffe-helper/config/config_env.go:
```diff
@@ -0,0 +1,62 @@
+		daemonMode = true
+	case "false", "0", "f", "no", "n", "off":
+		daemonMode = false
+	default:
```
can you add docuemntation about this?

faisal-memon replied:
Added README documentation for SPIFFE_HLP_JWT_SVIDS, including the YAML/JSON array format.

MarcosDY commented on cmd/spiffe-helper/config/config.go:
```diff
@@ -11,9 +12,11 @@ import (
 
 	"github.com/hashicorp/hcl"
 	"github.com/hashicorp/hcl/hcl/token"
+	"github.com/ilyakaznacheev/cleanenv"
```
why this dep is required? looks like it is not getting updates for some time, can we avoid it?

faisal-memon replied:
Swapped this from cleanenv to caarlos0/env in a7ff5be. That also let us use a single SPIFFE_HLP_ prefix and remove the manual daemon-mode/env parsing.

MarcosDY commented on script/docker/test-env-vars.sh:
```diff
@@ -0,0 +1,404 @@
+    fi
+    
+    # Check if the error is a configuration error (which is a failure) vs connection error (which is expected)
+    if echo "${output}" | grep -q "invalid configuration\|failed to parse"; then
```
I think we may be hiding errors here, right now there is no way to "dsable" config file since we have a default value, this test is not really using env vars and may be trying to use the file, can you verify if  log that confirms we are going into envvars code/

faisal-memon replied:
Updated this path so the default missing helper.conf loads from env vars, and fixed the Docker test script to validate the reconciled config from logs. I also made the script read the Go version from go.mod, handle multiline env vars safely, and run locally on macOS. make test-integration-docker now passes locally: 10 passed, 0 failed.

MarcosDY commented on cmd/spiffe-helper/config/config_test.go:
```diff
@@ -395,7 +530,577 @@ func TestDaemonModeFlag(t *testing.T) {
+		t.Run(tt.name, func(t *testing.T) {
+			// Set environment variables
+			for key, value := range tt.envVars {
+				os.Setenv(key, value)
```
NIT: can you remove the use of os.Setenv in favor of t.Setenv?

faisal-memon replied:
Fixed. The env-var tests now use t.Setenv instead of os.Setenv/os.Unsetenv cleanup.

eminwux commented:
Hi @faisal-memon let me know if you want me to address the comments. I now have availability to work on this. Thanks

faisal-memon commented:
> Hi @faisal-memon let me know if you want me to address the comments. I now have availability to work on this. Thanks

That would be great. main concern is cleanenv seems to be abandoned so we need to find an alternative.

eminwux commented:
Hi @faisal-memon I have added a new option in the [pros/cons table](https://docs.google.com/document/d/1U6lrDWULeZmlIIdbfGzEUU7q_B7AYO0yOAiOJXIEaoQ)
I've included two options: (1) https://github.com/caarlos0/env (2) implement it natively in spiffe-helper (could be a separate lib also). I will share it in Slack to get opinions.

faisal-memon commented:
> Hi @faisal-memon I have added a new option in the [pros/cons table](https://docs.google.com/document/d/1U6lrDWULeZmlIIdbfGzEUU7q_B7AYO0yOAiOJXIEaoQ) I've included two options: (1) https://github.com/caarlos0/env (2) implement it natively in spiffe-helper (could be a separate lib also). I will share it in Slack to get opinions.

the caarlos0 one seems ok to me.
